rat/model.py

```diff
@@ -1,272 +1,340 @@
-    device_first_in_group_indicators: Dict[str, jax.numpy.array]
-    compiled_model: types.ModuleType
-
-    def _constrain_and_transform_parameters(self, unconstrained_parameter_vector, pad=True):
-        jacobian_adjustment, parameters = self.compiled_model.constrain_parameters(unconstrained_parameter_vector, pad)
-        return jacobian_adjustment, self.compiled_model.transform_parameters(
-            self.device_data, self.device_subscript_indices, self.device_first_in_group_indicators, parameters
-        )
-
-    def _log_density(self, include_jacobian, unconstrained_parameter_vector):
-        # Evaluate model log density given model, data, subscripts and unconstrained parameters
-        jacobian_adjustment, parameters = self._constrain_and_transform_parameters(unconstrained_parameter_vector, pad=True)
-        target = self.compiled_model.evaluate_densities(self.device_data, self.device_subscript_indices, parameters)
-        return target + (jacobian_adjustment if include_jacobian else 0.0)
+class ExecuteException(AstException):
+    def __init__(self, message: str, node: ast.ModelBase):
+        super().__init__("evaluating log density", message, node)
 
-    def _prepare_draws_and_dfs(self, device_unconstrained_draws):
-        unconstrained_draws = numpy.array(device_unconstrained_draws)
-        num_draws = unconstrained_draws.shape[0]
-        num_chains = unconstrained_draws.shape[1]
-        constrained_draws = {}
-
-        constrain_and_transform_parameters_no_pad_jax = jax.jit(lambda x: self._constrain_and_transform_parameters(x, pad=False))
-        for draw in range(num_draws):
-            for chain in range(num_chains):
-                jacobian_adjustment, device_constrained_variables = constrain_and_transform_parameters_no_pad_jax(
-                    unconstrained_draws[draw, chain]
-                )
-
-                for name, device_constrained_variable in device_constrained_variables.items():
-                    if name not in constrained_draws:
-                        constrained_draws[name] = numpy.zeros((num_draws, num_chains) + device_constrained_variable.shape)
-                    constrained_draws[name][draw, chain] = numpy.array(device_constrained_variable)
-
-        # Copy back to numpy arrays
-        return constrained_draws, self.base_df_dict
-
-    def __init__(
-        self,
-        data: Union[pandas.DataFrame, Dict[str, pandas.DataFrame]],
-        model_string: str = None,
-        parsed_lines: List[ast.Expr] = None,
-        compile_path: str = None,
-        overwrite: bool = False,
-    ):
-        """
-        Create a model from some data (`data`) and a model (specified as a string, `model_string`).
 
-        If compile_path is not None, then write the compiled model to the given path (will only overwrite
-        existing files if the overwrite flag is true)
+class CodeExecutor(NodeWalker):
+    arguments: Dict[ast.Variable, Any]
+    parameters: Dict[str, Any]
```

Comment:
I would also appreciate a more verbose name, such as `parameter_value_arrays`

Reply:
We could split these by autodiff/non-autodiff or something? I kinda like parameters here. I think I used parameters just cause that's what I called it in other places and that's what it was in the codegen version before.

Reply:
Maybe the naming should be Compiletime vs. Runtime, or Constant vs. Dynamic (and we already use the constant/dynamic verbage on the variable table).

The data/parameter split is awkward here because transformed parameters are dynamic without being sampled parameters.

Reply:
Yup - I like the compile-time vs runtime separation. I'm not sure whether `CodeExecutor` would be the place to nuance this separation since this is just a generic executor class.

Reply:
No need to worry if this is perfect place or not. If it's easier to read it's easier to read. Can always change the naming later.